### PR TITLE
elided lifetime

### LINKED ISSUE #45992: lint against "silent elision" in structs/enums/etc
Now that support for `'_` has landed, as part of https://github.com/rust-lang/rust/issues/44524, it would be good to start linting for cases where it ought to be used. In particular, we wish to warn if there is a struct with lifetime parameters that are elided:

```rust
struct Foo<'a> { x: &'a u32 }
fn foo(x: &Foo) {
//         ^^^ warning: hidden lifetime parameters are deprecated, try `Foo<'_>`
}
```

This should be fairly straightforward to do during `resolve_lifetimes`.

### PR DESCRIPTION
**Pull request by Dylan-DPC-zz**
Closes #45992 

Hey
Having a problem with my config so decided to make a WIP PR nevertheless. Will add some more tests.

### CODE REVIEW

**Comment by michaelwoerister** [comment]
Unassigning myself since this isn't really my area. Please make sure to make the appropriate folks from @rust-lang/compiler aware of your PR, @Dylan-DPC.

**Comment by Dylan-DPC** [comment]
@nikomatsakis

**Comment by eddyb** [comment]
BTW you should rebase instead of merging.

**Comment by Dylan-DPC** [comment]
Yeah thanks. But i resolved the conflicts on the Github UI (since it were minor conflicts) which i assume merges by default.

**Comment by eddyb** [comment]
You can still rebase, and it will remove the merge commit.

**Comment by Dylan-DPC** [comment]
Yeah sure

**Comment by nikomatsakis** [comment]
r? @nikomatsakis

Sorry, somehow this didn't wind up assigned to me.

**Comment by nikomatsakis** [comment]
@Dylan-DPC so it seems like you're not actually *issuing* the lint anywhere -- is that right? Do you need help with just how to do this?

**Comment by nikomatsakis** [comment]
Ah, actually, you're probably going to hit the same issue that @gaurikholkar hit in https://github.com/rust-lang/rust/pull/46441. In particular, we can't issue lints from `resolve_lifetimes` right now because of when it executes. I fixed this as part of [this commit](https://github.com/rust-lang/rust/pull/46441/commits/15fa4ef94ecf6e0bbf3c3cb9e1c412ea47bd49ca) in Gauri's PR. You should probably rebase atop that commit, but maybe I should factor it out into a separate PR so you are not blocked on https://github.com/rust-lang/rust/pull/46441 having landed.

**Comment by nikomatsakis** [comment]
That said, https://github.com/rust-lang/rust/pull/46441/ is a good PR to look at for how to issue lints in general. =)

**Comment by Dylan-DPC** [comment]
Yeah i will rebase ontop of it.

**Comment by aidanhs** [comment]
Hi @Dylan-DPC, how are you doing with this? https://github.com/rust-lang/rust/pull/46657 is merged so I think you can rebase on top of that now.

File: src/librustc/lint/builtin.rs [hunk 1]
```diff
@@ -292,6 +298,7 @@ impl LintPass for HardwiredLints {
             COERCE_NEVER,
             SINGLE_USE_LIFETIME,
             TYVAR_BEHIND_RAW_POINTER
+            ELIDED_LIFETIME_IN_PATH
```
Comment (gaurikholkar) [inline src/librustc/lint/builtin.rs]: Missing a comma on the line above

**Comment by Dylan-DPC** [comment]
hi @nikomatsakis 
> so it seems like you're not actually issuing the lint anywhere -- is that right? Do you need help with just how to do this?

yeah trying to figure out how to issue. Should i just issue it when deprecated is true?

File: src/test/ui/in-band-lifetimes/ellided_lifetimes.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
```
Comment (gaurikholkar) [inline src/test/ui/in-band-lifetimes/ellided_lifetimes.rs]: Add the .stderr file for the ui test

File: src/librustc/lint/builtin.rs (hunk 1 above)
Comment (Dylan-DPC) [reply src/librustc/lint/builtin.rs]: Thanks fixed

File: src/test/ui/in-band-lifetimes/ellided_lifetimes.rs (hunk 2 above)
Comment (gaurikholkar) [reply src/test/ui/in-band-lifetimes/ellided_lifetimes.rs]: You can run ./x.py test --stage 1 src/test/ui to get the stderr file

Comment (Dylan-DPC) [reply src/test/ui/in-band-lifetimes/ellided_lifetimes.rs]: Ya thanks. Just have to work on some stuff before that

File: src/test/ui/in-band-lifetimes/ellided_lifetimes.rs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+#![deny(elided_lifetime_in_path)]
+struct Foo<'a> { x: &'a u32 }
+fn foo(x: &Foo) {
+//         ^^^ warning: hidden lifetime parameters are deprecated, try `Foo<'_>`
```
Comment (nikomatsakis) [inline src/test/ui/in-band-lifetimes/ellided_lifetimes.rs]: This should be `//~^ ERROR: ...`

That is an annotation that indicates that we expect an error on the line above. The `.stderr` file will give more precise information.

**Review by nikomatsakis — COMMENTED** [review]
Obviously you have to add the code that actually issues the lint =) let me know if you need any tips on that

**Comment by nikomatsakis** [comment]
Oh, I missed this comment:

> yeah trying to figure out how to issue. Should i just issue it when deprecated is true?

Yep, I think so.

**Comment by nikomatsakis** [comment]
To issue the lint (i.e., when deprecated is true), you would want something like this:

https://github.com/rust-lang/rust/blob/master/src/librustc/middle/resolve_lifetime.rs#L1228-L1234

**Comment by Dylan-DPC** [comment]
Thanks. My question was do i need to do perform any other check?

**Comment by nikomatsakis** [comment]
@Dylan-DPC I don't *think* so -- I think that would suffice. But I guess we should make some tests of various scenarios to be sure.

**Comment by Dylan-DPC** [comment]
Hey @nikomatsakis 
I have to create a `LifetimeContext` for this right? in order to call `struct_span_lint_node`?

File: src/librustc/middle/resolve_lifetime.rs [hunk 4]
```diff
@@ -1803,7 +1803,7 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
         }
     }
 
-    fn resolve_elided_lifetimes(&mut self, lifetime_refs: &'tcx [hir::Lifetime]) {
+    fn resolve_elided_lifetimes(&mut self, lifetime_refs: &'tcx [hir::Lifetime], deprecated: bool) {
```
Comment (nikomatsakis) [inline src/librustc/middle/resolve_lifetime.rs]: @Dylan-DPC 

Here is an example of adding a lint from the code in question:

https://github.com/rust-lang/rust/blob/fdc18b3067b5bad257ccbe7400e3c4fb617e9e18/src/librustc/middle/resolve_lifetime.rs#L1228-L1234

I think you should be able to do something similar from this function here when `deprecated` is true.

Comment (Dylan-DPC) [reply src/librustc/middle/resolve_lifetime.rs]: Hey Thanks
I have followed that code already was wondering if i had to create the `LifetimeContext` object just like what is being done in the function. I guess i have to.

Comment (nikomatsakis) [reply src/librustc/middle/resolve_lifetime.rs]: I don' think you have to do that, no. I am confused. You should have a `LifetimeContext` available. 

Basically, that function does some things in addition to creating the lint -- that is why it makes the new `LifetimeContext` struct -- you are not doing those things. The only thing you should need is the call to `struct_span_lint_node`

Comment (nikomatsakis) [reply src/librustc/middle/resolve_lifetime.rs]: > I have followed that code already

What do you mean by "followed" here? I don't see (in the PR) any calls to `struct_span_lint_node`, maybe you forgot to push a commit?

Comment (nikomatsakis) [reply src/librustc/middle/resolve_lifetime.rs]: You should not have to create a `LifetimeContext`, no. That function does a few other things in addition to issuing the lint -- in particular, it pushes a "subscope" -- and that is why it creates a `LifetimeContext`. You can just use the one you have, I expect.

Comment (Dylan-DPC) [reply src/librustc/middle/resolve_lifetime.rs]: > What do you mean by "followed" here? I don't see (in the PR) any calls to struct_span_lint_node, maybe you forgot to push a commit?

Ah I meant the code you had mentioned earlier. 

My question is: do I have access to this.tcx directly? I thought I didn't

**Comment by Dylan-DPC** [comment]
Looks like the `outer forbid lint` is blocking me in some way.

Stderr:
```
stderr:
------------------------------------------
{"message":"allow(unused_variables) overruled by outer forbid(unused)","code":{"code":"E0453","explanation":"\nA lint check attribute was overruled by a `forbid` directive set as an\nattribute on an enclosing scope, or on the command line with the `-F` option.\n\nExample of erroneous code:\n\n```compile_fail,E0453\n#![forbid(non_snake_case)]\n\n#[allow(non_snake_case)]\nfn main() {\n    let MyNumber = 2; // error: allow(non_snake_case) overruled by outer\n                      //        forbid(non_snake_case)\n}\n```\n\nThe `forbid` lint setting, like `deny`, turns the corresponding compiler\nwarning into a hard error. Unlike `deny`, `forbid` prevents itself from being\noverridden by inner attributes.\n\nIf you're sure you want to override the lint check, you can change `forbid` to\n`deny` (or use `-D` instead of `-F` if the `forbid` setting was given as a\ncommand-line option) to allow the inner lint check attribute:\n\n```\n#![deny(non_snake_case)]\n\n#[allow(non_snake_case)]\nfn main() {\n    let MyNumber = 2; // ok!\n}\n```\n\nOtherwise, edit the code to pass the lint check, and remove the overruled\nattribute:\n\n```\n#![forbid(non_snake_case)]\n\nfn main() {\n    let my_number = 2;\n}\n```\n"},"level":"error","spans":[{"file_name":"/Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs","byte_start":837,"byte_end":853,"line_start":19,"line_end":19,"column_start":9,"column_end":25,"is_primary":true,"text":[{"text":"#[allow(unused_variables)] //~ ERROR overruled","highlight_start":9,"highlight_end":25}],"label":"overruled by previous forbid","suggested_replacement":null,"expansion":null},{"file_name":"/Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs","byte_start":803,"byte_end":809,"line_start":17,"line_end":17,"column_start":11,"column_end":17,"is_primary":false,"text":[{"text":"#![forbid(unused, non_snake_case)]","highlight_start":11,"highlight_end":17}],"label":"`forbid` level set here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0453]: allow(unused_variables) overruled by outer forbid(unused)\n  --> /Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs:19:9\n   |\n17 | #![forbid(unused, non_snake_case)]\n   |           ------ `forbid` level set here\n18 | \n19 | #[allow(unused_variables)] //~ ERROR overruled\n   |         ^^^^^^^^^^^^^^^^ overruled by previous forbid\n\n"}
{"message":"allow(unused) overruled by outer forbid(unused)","code":{"code":"E0453","explanation":"\nA lint check attribute was overruled by a `forbid` directive set as an\nattribute on an enclosing scope, or on the command line with the `-F` option.\n\nExample of erroneous code:\n\n```compile_fail,E0453\n#![forbid(non_snake_case)]\n\n#[allow(non_snake_case)]\nfn main() {\n    let MyNumber = 2; // error: allow(non_snake_case) overruled by outer\n                      //        forbid(non_snake_case)\n}\n```\n\nThe `forbid` lint setting, like `deny`, turns the corresponding compiler\nwarning into a hard error. Unlike `deny`, `forbid` prevents itself from being\noverridden by inner attributes.\n\nIf you're sure you want to override the lint check, you can change `forbid` to\n`deny` (or use `-D` instead of `-F` if the `forbid` setting was given as a\ncommand-line option) to allow the inner lint check attribute:\n\n```\n#![deny(non_snake_case)]\n\n#[allow(non_snake_case)]\nfn main() {\n    let MyNumber = 2; // ok!\n}\n```\n\nOtherwise, edit the code to pass the lint check, and remove the overruled\nattribute:\n\n```\n#![forbid(non_snake_case)]\n\nfn main() {\n    let my_number = 2;\n}\n```\n"},"level":"error","spans":[{"file_name":"/Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs","byte_start":897,"byte_end":903,"line_start":22,"line_end":22,"column_start":9,"column_end":15,"is_primary":true,"text":[{"text":"#[allow(unused)] //~ ERROR overruled","highlight_start":9,"highlight_end":15}],"label":"overruled by previous forbid","suggested_replacement":null,"expansion":null},{"file_name":"/Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs","byte_start":803,"byte_end":809,"line_start":17,"line_end":17,"column_start":11,"column_end":17,"is_primary":false,"text":[{"text":"#![forbid(unused, non_snake_case)]","highlight_start":11,"highlight_end":17}],"label":"`forbid` level set here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0453]: allow(unused) overruled by outer forbid(unused)\n  --> /Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs:22:9\n   |\n17 | #![forbid(unused, non_snake_case)]\n   |           ------ `forbid` level set here\n...\n22 | #[allow(unused)] //~ ERROR overruled\n   |         ^^^^^^ overruled by previous forbid\n\n"}
{"message":"allow(bad_style) overruled by outer forbid(non_snake_case)","code":{"code":"E0453","explanation":"\nA lint check attribute was overruled by a `forbid` directive set as an\nattribute on an enclosing scope, or on the command line with the `-F` option.\n\nExample of erroneous code:\n\n```compile_fail,E0453\n#![forbid(non_snake_case)]\n\n#[allow(non_snake_case)]\nfn main() {\n    let MyNumber = 2; // error: allow(non_snake_case) overruled by outer\n                      //        forbid(non_snake_case)\n}\n```\n\nThe `forbid` lint setting, like `deny`, turns the corresponding compiler\nwarning into a hard error. Unlike `deny`, `forbid` prevents itself from being\noverridden by inner attributes.\n\nIf you're sure you want to override the lint check, you can change `forbid` to\n`deny` (or use `-D` instead of `-F` if the `forbid` setting was given as a\ncommand-line option) to allow the inner lint check attribute:\n\n```\n#![deny(non_snake_case)]\n\n#[allow(non_snake_case)]\nfn main() {\n    let MyNumber = 2; // ok!\n}\n```\n\nOtherwise, edit the code to pass the lint check, and remove the overruled\nattribute:\n\n```\n#![forbid(non_snake_case)]\n\nfn main() {\n    let my_number = 2;\n}\n```\n"},"level":"error","spans":[{"file_name":"/Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs","byte_start":947,"byte_end":956,"line_start":25,"line_end":25,"column_start":9,"column_end":18,"is_primary":true,"text":[{"text":"#[allow(bad_style)] //~ ERROR overruled","highlight_start":9,"highlight_end":18}],"label":"overruled by previous forbid","suggested_replacement":null,"expansion":null},{"file_name":"/Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs","byte_start":811,"byte_end":825,"line_start":17,"line_end":17,"column_start":19,"column_end":33,"is_primary":false,"text":[{"text":"#![forbid(unused, non_snake_case)]","highlight_start":19,"highlight_end":33}],"label":"`forbid` level set here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0453]: allow(bad_style) overruled by outer forbid(non_snake_case)\n  --> /Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs:25:9\n   |\n17 | #![forbid(unused, non_snake_case)]\n   |                   -------------- `forbid` level set here\n...\n25 | #[allow(bad_style)] //~ ERROR overruled\n   |         ^^^^^^^^^ overruled by previous forbid\n\n"}
{"message":"aborting due to 6 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 6 previous errors\n\n"}

------------------------------------------

thread '[ui] ui/lint/outer-forbid.rs' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:2776:8
```

**Comment by Dylan-DPC** [comment]
Bumped into this problem:
```
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
{"message":"allow(unused_variables) overruled by outer forbid(unused)","code":{"code":"E0453","explanation":"\nA lint check attribute was overruled by a `forbid` directive set as an\nattribute on an enclosing scope, or on the command line with the `-F` option.\n\nExample of erroneous code:\n\n```compile_fail,E0453\n#![forbid(non_snake_case)]\n\n#[allow(non_snake_case)]\nfn main() {\n    let MyNumber = 2; // error: allow(non_snake_case) overruled by outer\n                      //        forbid(non_snake_case)\n}\n```\n\nThe `forbid` lint setting, like `deny`, turns the corresponding compiler\nwarning into a hard error. Unlike `deny`, `forbid` prevents itself from being\noverridden by inner attributes.\n\nIf you're sure you want to override the lint check, you can change `forbid` to\n`deny` (or use `-D` instead of `-F` if the `forbid` setting was given as a\ncommand-line option) to allow the inner lint check attribute:\n\n```\n#![deny(non_snake_case)]\n\n#[allow(non_snake_case)]\nfn main() {\n    let MyNumber = 2; // ok!\n}\n```\n\nOtherwise, edit the code to pass the lint check, and remove the overruled\nattribute:\n\n```\n#![forbid(non_snake_case)]\n\nfn main() {\n    let my_number = 2;\n}\n```\n"},"level":"error","spans":[{"file_name":"/Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs","byte_start":837,"byte_end":853,"line_start":19,"line_end":19,"column_start":9,"column_end":25,"is_primary":true,"text":[{"text":"#[allow(unused_variables)] //~ ERROR overruled","highlight_start":9,"highlight_end":25}],"label":"overruled by previous forbid","suggested_replacement":null,"expansion":null},{"file_name":"/Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs","byte_start":803,"byte_end":809,"line_start":17,"line_end":17,"column_start":11,"column_end":17,"is_primary":false,"text":[{"text":"#![forbid(unused, non_snake_case)]","highlight_start":11,"highlight_end":17}],"label":"`forbid` level set here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0453]: allow(unused_variables) overruled by outer forbid(unused)\n  --> /Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs:19:9\n   |\n17 | #![forbid(unused, non_snake_case)]\n   |           ------ `forbid` level set here\n18 | \n19 | #[allow(unused_variables)] //~ ERROR overruled\n   |         ^^^^^^^^^^^^^^^^ overruled by previous forbid\n\n"}
{"message":"allow(unused) overruled by outer forbid(unused)","code":{"code":"E0453","explanation":"\nA lint check attribute was overruled by a `forbid` directive set as an\nattribute on an enclosing scope, or on the command line with the `-F` option.\n\nExample of erroneous code:\n\n```compile_fail,E0453\n#![forbid(non_snake_case)]\n\n#[allow(non_snake_case)]\nfn main() {\n    let MyNumber = 2; // error: allow(non_snake_case) overruled by outer\n                      //        forbid(non_snake_case)\n}\n```\n\nThe `forbid` lint setting, like `deny`, turns the corresponding compiler\nwarning into a hard error. Unlike `deny`, `forbid` prevents itself from being\noverridden by inner attributes.\n\nIf you're sure you want to override the lint check, you can change `forbid` to\n`deny` (or use `-D` instead of `-F` if the `forbid` setting was given as a\ncommand-line option) to allow the inner lint check attribute:\n\n```\n#![deny(non_snake_case)]\n\n#[allow(non_snake_case)]\nfn main() {\n    let MyNumber = 2; // ok!\n}\n```\n\nOtherwise, edit the code to pass the lint check, and remove the overruled\nattribute:\n\n```\n#![forbid(non_snake_case)]\n\nfn main() {\n    let my_number = 2;\n}\n```\n"},"level":"error","spans":[{"file_name":"/Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs","byte_start":897,"byte_end":903,"line_start":22,"line_end":22,"column_start":9,"column_end":15,"is_primary":true,"text":[{"text":"#[allow(unused)] //~ ERROR overruled","highlight_start":9,"highlight_end":15}],"label":"overruled by previous forbid","suggested_replacement":null,"expansion":null},{"file_name":"/Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs","byte_start":803,"byte_end":809,"line_start":17,"line_end":17,"column_start":11,"column_end":17,"is_primary":false,"text":[{"text":"#![forbid(unused, non_snake_case)]","highlight_start":11,"highlight_end":17}],"label":"`forbid` level set here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0453]: allow(unused) overruled by outer forbid(unused)\n  --> /Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs:22:9\n   |\n17 | #![forbid(unused, non_snake_case)]\n   |           ------ `forbid` level set here\n...\n22 | #[allow(unused)] //~ ERROR overruled\n   |         ^^^^^^ overruled by previous forbid\n\n"}
{"message":"allow(bad_style) overruled by outer forbid(non_snake_case)","code":{"code":"E0453","explanation":"\nA lint check attribute was overruled by a `forbid` directive set as an\nattribute on an enclosing scope, or on the command line with the `-F` option.\n\nExample of erroneous code:\n\n```compile_fail,E0453\n#![forbid(non_snake_case)]\n\n#[allow(non_snake_case)]\nfn main() {\n    let MyNumber = 2; // error: allow(non_snake_case) overruled by outer\n                      //        forbid(non_snake_case)\n}\n```\n\nThe `forbid` lint setting, like `deny`, turns the corresponding compiler\nwarning into a hard error. Unlike `deny`, `forbid` prevents itself from being\noverridden by inner attributes.\n\nIf you're sure you want to override the lint check, you can change `forbid` to\n`deny` (or use `-D` instead of `-F` if the `forbid` setting was given as a\ncommand-line option) to allow the inner lint check attribute:\n\n```\n#![deny(non_snake_case)]\n\n#[allow(non_snake_case)]\nfn main() {\n    let MyNumber = 2; // ok!\n}\n```\n\nOtherwise, edit the code to pass the lint check, and remove the overruled\nattribute:\n\n```\n#![forbid(non_snake_case)]\n\nfn main() {\n    let my_number = 2;\n}\n```\n"},"level":"error","spans":[{"file_name":"/Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs","byte_start":947,"byte_end":956,"line_start":25,"line_end":25,"column_start":9,"column_end":18,"is_primary":true,"text":[{"text":"#[allow(bad_style)] //~ ERROR overruled","highlight_start":9,"highlight_end":18}],"label":"overruled by previous forbid","suggested_replacement":null,"expansion":null},{"file_name":"/Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs","byte_start":811,"byte_end":825,"line_start":17,"line_end":17,"column_start":19,"column_end":33,"is_primary":false,"text":[{"text":"#![forbid(unused, non_snake_case)]","highlight_start":19,"highlight_end":33}],"label":"`forbid` level set here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0453]: allow(bad_style) overruled by outer forbid(non_snake_case)\n  --> /Users/dpc/Code/rust/src/test/ui/lint/outer-forbid.rs:25:9\n   |\n17 | #![forbid(unused, non_snake_case)]\n   |                   -------------- `forbid` level set here\n...\n25 | #[allow(bad_style)] //~ ERROR overruled\n   |         ^^^^^^^^^ overruled by previous forbid\n\n"}
{"message":"aborting due to 6 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 6 previous errors\n\n"}
```

**Comment by nikomatsakis** [comment]
@Dylan-DPC ok, so, to start, you [need a trailing newline](https://travis-ci.org/rust-lang/rust/builds/332450708#L1818), according to travis.

**Comment by nikomatsakis** [comment]
Second, your source is generating a number of other warnings. I think these are mostly related to the name `x`, which is unused, but also `Foo`, which is dead code.

I would add `#![allow(unused_variables, dead_code)]` to the top of the test, I think.

**Comment by Dylan-DPC** [comment]
Hey thanks. Ye I knew what was causing it but it didn't work when I tried to fix it.

File: src/test/ui/in-band-lifetimes/ellided_lifetimes.rs [hunk 5]
```diff
@@ -7,10 +7,13 @@
...
 // except according to those terms.
+#![allow(warnings)]
+#![allow(unused_variables, dead_code, unused, bad_style)]
 #![deny(elided_lifetime_in_path)]
+
 struct Foo<'a> { x: &'a u32 }
-fn foo(x: &Foo) {
-         //~^ ERROR: hidden lifetime parameters are deprecated, try `Foo<'_>`
-}
+    fn foo(x: &Foo) {
```
Comment (nikomatsakis) [inline src/test/ui/in-band-lifetimes/ellided_lifetimes.rs]: why the indentation here?

File: src/test/ui/in-band-lifetimes/ellided_lifetimes.rs [hunk 6]
```diff
@@ -7,10 +7,13 @@
 // <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
+#![allow(warnings)]
+#![allow(unused_variables, dead_code, unused, bad_style)]
```
Comment (nikomatsakis) [inline src/test/ui/in-band-lifetimes/ellided_lifetimes.rs]: also, you shouldn't need *both* allow warnings *and* the individual lints -- one ought to suffice, I would think

Comment (Dylan-DPC) [reply src/test/ui/in-band-lifetimes/ellided_lifetimes.rs]: True. I thought it might work with allow warnings but it doesn't for some reason 🤔

**Comment by Dylan-DPC** [comment]
Just did a rebase and I now get the lint message in the stderr: 
> "message":"hidden lifetime parameters are deprecated, try `Foo<'_>`","code":{"code":"elided_lifetime_in_path","explanation":null},  
( haven't pushed that code yet, will do it today evening)

**Comment by nikomatsakis** [comment]
@Dylan-DPC not sure I understand, but once you push I will take a look :)

**Comment by Dylan-DPC** [comment]
The code is already updated. You can check the Travis CI output you will understand what i'm trying to say.

**Comment by nikomatsakis** [comment]
Ah, ok

**Comment by nikomatsakis** [comment]
@Dylan-DPC did you run the script to update the UI reference files?

See [this line of the output for instructions](https://travis-ci.org/rust-lang/rust/builds/332910551#L1612)

**Comment by nikomatsakis** [comment]
@Dylan-DPC somehow you altered [another test too](https://travis-ci.org/rust-lang/rust/builds/333844120#L2319)

I often run `./x.py test --stage 1 src/test/ui` a few times, also running `update-all-references.sh`, until it comes to a "fixed point". This is kind of annoying, there are a few bugs in how the harness works such that sometimes (e.g., when it skips a test) you have outdated ref files lying around. We should fix that sometime.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@Dylan-DPC -- actually, do you want to squash the commits? Code looks good to me =)

**Comment by nikomatsakis** [comment]
r=me once squashed

**Comment by estebank** [comment]
[The merge is removing stuff from `RELEASES.md`](https://github.com/rust-lang/rust/pull/46254/commits/e376b2039e40af5f4bf382b445ad6f773f5c7d4a).

**Comment by Dylan-DPC** [comment]
Ah thanks. Give me a minute will fix this

**Comment by Dylan-DPC** [comment]
@estebank what should be my next step? Should i rebase with master? Cause I see a huge number of commits being the difference

**Comment by estebank** [comment]
Whenever possible, I avoid merges and rebase against latest master.

**Comment by Dylan-DPC** [comment]
Ah shucks, messed up with the branch. I guess submitting a new PR will be a better idea 😅

**Comment by Dylan-DPC** [comment]
r=@nikomatsakis

**Comment by estebank** [comment]
```
[00:04:06] tidy error: /checkout/src/test/ui/in-band-lifetimes/ellided-lifetimes.rs: incorrect license
[00:04:06] tidy error: /checkout/src/test/ui/in-band-lifetimes/ellided-lifetimes.rs: missing trailing newline
[00:04:07] some tidy checks failed
[00:04:07] 
```

**Comment by estebank** [comment]
```
[00:07:17]    Compiling syntax_ext v0.0.0 (file:///checkout/src/libsyntax_ext)
[00:07:23] error: expected item, found `,`
[00:07:23]    --> librustc/lint/builtin.rs:247:2
[00:07:23]     |
[00:07:23] 247 | },
[00:07:23]     |  ^
[00:07:23] 
[00:07:23] error: aborting due to previous error
[00:07:23] 
[00:07:23] error: Could not compile `rustc`.
```

File: src/librustc/lint/builtin.rs [hunk 7]
```diff
@@ -244,6 +244,12 @@ declare_lint! {
     pub TYVAR_BEHIND_RAW_POINTER,
     Warn,
     "raw pointer to an inference variable"
+},
+
++declare_lint! {
```
Comment (estebank) [inline src/librustc/lint/builtin.rs]: is the leading + correct?

File: src/librustc/lint/builtin.rs [hunk 8]
```diff
@@ -244,6 +244,12 @@ declare_lint! {
     pub TYVAR_BEHIND_RAW_POINTER,
     Warn,
     "raw pointer to an inference variable"
+},
```
Comment (estebank) [inline src/librustc/lint/builtin.rs]: is this comma supposed to be here?

**Review by estebank — COMMENTED** [review]
Once you've fixed the compile error, squash the commits into one.

**Comment by Dylan-DPC** [comment]
@estebank done

**Comment by estebank** [comment]
```

[00:41:24] failures:
[00:41:24] 
[00:41:24] ---- [ui] ui/in-band-lifetimes/ellided-lifetimes.rs stdout ----
[00:41:24] 	diff of stderr:
[00:41:24] 
[00:41:24] 1	error: hidden lifetime parameters are deprecated, try `Foo<'_>`
[00:41:24] +	  --> $DIR/ellided-lifetimes.rs:15:12
[00:41:24] -	  --> $DIR/ellided_lifetimes.rs:15:16
[00:41:24] 3	   |
[00:41:24] +	15 | fn foo(x: &Foo) {
[00:41:24] +	   |            ^^^
[00:41:24] -	15 |     fn foo(x: &Foo) {
[00:41:24] -	   |                ^^^
[00:41:24] 6	   |
[00:41:24] 7	note: lint level defined here
[00:41:24] +	  --> $DIR/ellided-lifetimes.rs:12:9
[00:41:24] -	  --> $DIR/ellided_lifetimes.rs:12:9
[00:41:24] 9	   |
[00:41:24] 10	12 | #![deny(elided_lifetime_in_path)]
[00:41:24] 11	   |         ^^^^^^^^^^^^^^^^^^^^^^^
[00:41:24] 
[00:41:24] 12	
[00:41:24] 13	error: aborting due to previous error
[00:41:24] +	
[00:41:24] +	
[00:41:24] 
```

**Comment by Dylan-DPC** [comment]
@estebank so it means that it is failing due to formatting and line number being different? can i edit the .stderr file to match that?

**Comment by estebank** [comment]
Correct. Fixing the `.stderr` file for all the diffs shown above (including the file name change from using `_` to `-`) and the test _should_ pass.

**Comment by estebank** [comment]
Also, looking at the commit list I see a merge commit that shouldn't be there. Only your commit should be there.

**Comment by estebank** [comment]
```
[00:43:44] ---- [ui] ui/in-band-lifetimes/ellided-lifetimes.rs stdout ----
[00:43:44] 	diff of stderr:
[00:43:44] 
[00:43:44] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:476:22
[00:43:44] 2	  --> $DIR/ellided-lifetimes.rs:15:12
[00:43:44] 3	   |
[00:43:44] 4	15 | fn foo(x: &Foo) {
[00:43:44] +	   |            ^^^
[00:43:44] -	   |                ^^^
[00:43:44] 6	   |
[00:43:44] 7	note: lint level defined here
[00:43:44] 8	  --> $DIR/ellided-lifetimes.rs:12:9
[00:43:44] 
[00:43:44] 9	   |
[00:43:44] +	12 | #![deny(elided_lifetime_in_path)]
[00:43:44] -	12 | #![deny(elided-lifetime_in_path)]
[00:43:44] 11	   |         ^^^^^^^^^^^^^^^^^^^^^^^
[00:43:44] 12	
[00:43:44] 13	error: aborting due to previous error
```

Also, remember to squash your commits and rebase against latest master.

**Comment by Dylan-DPC** [comment]
Something is odd here. Did a rebase with master, but shouldn't have got those commits 🤔

**Comment by estebank** [comment]
From the result, it seems to me that while in `master`, you did `git rebase your-branch`, instead of being in `your-branch` and doing `git rebase master`.

**Comment by nikomatsakis** [comment]
Ah, git...

**Comment by Dylan-DPC** [comment]
I went through the commands. Don't see anything that I did wrong. Anyway, here we go again.

**Comment by nikomatsakis** [comment]
@Dylan-DPC if you want, I can take a crack at it =) 

But normally you would do this:

```
git fetch rust-lang # or origin, or whatever
```

**Comment by nikomatsakis** [comment]
I pretty much *never* rebase without `-i`.

**Comment by Dylan-DPC** [comment]
Looks like finally success 😄 

> I pretty much never rebase without -i.
Wow. I usually avoid interactive rebases because of the fact that you can accidentally lose commits.

**Comment by eddyb** [comment]
@Dylan-DPC `git reflog` means you only lose commits if you don't look at them for a month or so.

**Comment by shepmaster** [comment]
> The job exceeded the maximum time limit for jobs, and has been terminated.

@bors retry

**Pull request merged**